src/main.rs

```diff
@@ -1,189 +1,169 @@
+            }
+        }
+    }
+
+    impl NetworkBehaviourEventProcess<KademliaEvent> for MyBehaviour {
+
+        // Called when `kademlia` produces an event.
+        fn inject_event(&mut self, message: KademliaEvent) {
+            match message {
+                KademliaEvent::OutboundQueryCompleted { result, .. } => match result {
+
+                    QueryResult::GetRecord(Ok(ok)) => {
+                        for PeerRecord {
+                            record: Record { key, value, .. },
+                            ..
+                        } in ok.records
+                        {
+                            println!(
+                                "Value for {:?} is:- {:?}",
+                                std::str::from_utf8(key.as_ref()).unwrap(),
+                                std::str::from_utf8(&value).unwrap(),
+                            );
+                        }
+                    }
+                    QueryResult::GetRecord(Err(err)) => {
+                        eprintln!("Failed to get record: {:?}", err);
+                    }
+                    QueryResult::PutRecord(Ok(PutRecordOk { key })) => {
+                        println!("Data successfully saved for {:?}", std::str::from_utf8(key.as_ref()).unwrap());
+                    }
+                    QueryResult::PutRecord(Err(err)) => {
+                        eprintln!("Failed to put record: {:?}", err);
+                    }
+
+                    _ => {}
+                },
+                _ => {}
+            }
+        }
+    }
 
     // Create a swarm to manage peers and events.
     let mut swarm = {
         // Create a Kademlia behaviour.
-        let store = MemoryStore::new(local_peer_id.clone());
-        let kademlia = Kademlia::new(local_peer_id.clone(), store);
-        let behaviour = MyBehaviour { kademlia};
+        let store = MemoryStore::new(local_peer_id);
+        let mut kad_config = KademliaConfig::default();
+        kad_config.set_connection_idle_timeout(Duration::from_secs(100000));
+        let kademlia = Kademlia::with_config(local_peer_id, store, kad_config);
+
+        let identify_config = IdentifyConfig::new("dht/1.0.0".to_string(),local_key.public().clone());
+        let identify = Identify::new(identify_config);
+        let behaviour = MyBehaviour { kademlia, identify };
+
         Swarm::new(transport, behaviour, local_peer_id)
     };
-    // Listen on all interfaces and whatever port the OS assigns.
-    &swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
-    // println!("swarm listening on /ip4/0.0.0.0/tcp/7865");
 
+    // Read full lines from stdin
     let mut stdin = io::BufReader::new(io::stdin()).lines();
-    let swarm_thread = start_swarm(&mut stdin, &mut swarm);
 
+    // Listen on all interfaces and whatever port the OS assigns.
+    swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
 
-    futures::join!(swarm_thread);
+    let str = "identity";
+    let key = Key::new(&str);
```

Comment:
I think this can be removed and the below would can be used in a single line
```
    swarm
        .behaviour_mut()
        .kademlia
        .start_providing(Key::new(&"identity"))
        .unwrap();
```

Reply:
I think you missed this one.